src/parcsr_ls/par_relax.c

```diff
@@ -16,4395 +16,1544 @@
-                       for (jj = A_diag_i[i]+1; jj < A_diag_i[i+1]; jj++)
-                       {
-                          ii = A_diag_j[jj];
-                          res0 -= A_diag_data[jj] * u_data[ii];
-                          res2 += A_diag_data[jj] * Vtemp_data[ii];
-                       }
-                       for (jj = A_offd_i[i]; jj < A_offd_i[i+1]; jj++)
-                       {
-                          ii = A_offd_j[jj];
-                          res -= A_offd_data[jj] * Vext_data[ii];
-                       }
-                       u_data[i] *= prod;
-                       u_data[i] += relax_weight*(omega*res + res0 +
-                             one_minus_omega*res2) / A_diag_data[A_diag_i[i]];
-                       /*u_data[i] += omega*(relax_weight*res + res0 +
-                         one_minus_weight*res2) / A_diag_data[A_diag_i[i]];*/
-                    }
-                 }
-              }
-           }
-        }
-#ifndef HYPRE_USING_PERSISTENT_COMM
-        if (num_procs > 1)
-        {
-           hypre_TFree(Vext_data, HYPRE_MEMORY_HOST);
-           hypre_TFree(v_buf_data, HYPRE_MEMORY_HOST);
-        }
-#endif
-#ifdef HYPRE_PROFILE
-        hypre_profile_times[HYPRE_TIMER_ID_RELAX] += hypre_MPI_Wtime();
-#endif
+      for (i = 0; i < num_rows; i++)
+      {
+         Ztemp_data[i] = u_data[i];
       }
-      break;
+      */
 
-      case 1: /* Gauss-Seidel VERY SLOW */
+#ifdef HYPRE_USING_OPENMP
+#pragma omp parallel for private(i,ii,j,jj,ns,ne,res,rest,size,sweep) HYPRE_SMP_SCHEDULE
+#endif
+      for (j = 0; j < num_threads; j++)
       {
-         if (num_procs > 1)
+         size = num_rows / num_threads;
+         rest = num_rows - size * num_threads;
+         if (j < rest)
          {
-            num_sends = hypre_ParCSRCommPkgNumSends(comm_pkg);
-            num_recvs = hypre_ParCSRCommPkgNumRecvs(comm_pkg);
-
-            v_buf_data = hypre_CTAlloc(HYPRE_Real,
-                  hypre_ParCSRCommPkgSendMapStart(comm_pkg,  num_sends), HYPRE_MEMORY_HOST);
-
-            Vext_data = hypre_CTAlloc(HYPRE_Real, num_cols_offd, HYPRE_MEMORY_HOST);
+            ns = j * size + j;
+            ne = (j + 1) * size + j + 1;
+         }
+         else
+         {
+            ns = j * size + rest;
+            ne = (j + 1) * size + rest;
+         }
 
-            status  = hypre_CTAlloc(hypre_MPI_Status, num_recvs+num_sends, HYPRE_MEMORY_HOST);
-            requests= hypre_CTAlloc(hypre_MPI_Request,  num_recvs+num_sends, HYPRE_MEMORY_HOST);
+         for (sweep = 0; sweep < num_sweeps; sweep++)
+         {
+            hypre_Iterator iter;
+            hypre_IteratorFirst(&iter) = ns;
+            hypre_IteratorLast(&iter) = ne;
+            hypre_IteratorDirection(&iter) = num_sweeps == 1 ? gs_order : sweep == 0 ? 1 : -1;
 
-            if (num_cols_offd)
+            for ( i = hypre_IteratorBegin(&iter); i != hypre_IteratorEnd(&iter); i += hypre_IteratorStep(&iter) )
```

Comment:
Hi @liruipeng .  Any idea about the overhead of this?  I see that you use 'static inline', which should have the potential to be quite fast, depending on the compiler.  I just don't have experience with it.

Reply:
`inline` is a hint to compilers to expand the code instead of making function calls. If the compiler takes the hint, there should be no cost. Even if it doesn't I assume it should not be too much problem, since sparse solvers generally are not CPU-bound.

Reply:
I would still do some testing/timing, just to be sure.  Compilers don't always do what we'd like them to do and loop overhead can't always be discounted, even in sparse linear algebra.

Reply:
I realized the whole `hypre_Iterator` thing is not necessary, so I removed it.

Reply:
Okay, looks good @liruipeng .  BTW, although removing the iterator looks like a better way to go in this case, note that I am not against the idea of iterators in general, nor the use of 'static inline' functions to define them.  I only wanted to be cautious about potential performance implications, which may have been nothing in the end, as you argued.  Thanks!

Reply:
Thanks @rfalgout. There was actually performance implications, not in the `hypre_iterator` per se, but in the overall strategy of putting all cases together in one piece of code. So, I reverted some of my effort to re-make a bunch of separate `inline` functions for different cases put in a new file `par_relax.h`. The desired refactored `hypre_BoomerAMGRelaxHybridGaussSeidel_core` still keeps a nice shape. I've tested the correctness and the efficiency, which look fine. There was no noticeable performance loss, while in some cases it could be even slightly faster than before.